src/Tools/dotnet-monitor/Program.cs

```diff
@@ -59,6 +57,14 @@ class Program
                 Argument = new Argument<bool>(name: "metrics", defaultValue: true )
             };
 
+        private static Option ReversedServerAddress() =>
+            new Option(
+                alias: "--reversed-server-address",
```

Comment:
Elsewhere this is called the diagnostic monitor address (for example in the environment variable) so I recommend we continue to use that naming. This could be --monitor-address or --address?

Reply:
I want to establish a naming convention for parameters such that all of the reversed server parameters are understood to apply to the reversed server mode. For example, we could add other parameters for connection limits (e.g. --reversed-server-max-connections), timeouts for waiting for connections (e.g. --reversed-server-wait-timeout), and other parameters. I'm not particularly fond of "monitor-address" because when I read that I think of "this is the address to be monitored" (I guess you could argue a server "monitors" whatever endpoint it has for connections) and the process is already called "dotnet-monitor" so it's implied that everything is related to monitoring already.  "address" feels too generic and does not imply the reversed server mode; you could conceivably use "address" for the forward mode to if we ever let that be configurable.

I'm open to changing the naming, but also would push for it to not be overly simplified and to keep in mind future related parameterization.

Reply:
I'll continue to think about the naming. I'm not attached to the specific ones I suggested but I do worry that the term "reversed server" isn't currently planned to show up anywhere else in the overall experience or documentation. I know we have used the phrase internally, but if it shows up in the public facing name I anticipate customers won't find it intuitive. In order to understand 'reversed server' the customer would first need to know what the runtime diagnostic server is and that this term is a derivative from it. I don't think they will know either of those things.

One possibility is that because we haven't spoken much in public or in docs about "diagnostic server" there is probably room to change our planned terminology without causing too much confusion, though the window won't be open much longer. @josalem @shirhatti @wiktork @sywhang if you guys had thoughts?

One thought would be:
```
Runtime env var: DOTNET_MANAGEMENT_PORT=\foo\bar\mysocket
Tool: dotnet-monitor --dotnet-management-port=\foo\bar\mysocket
```
In both contexts I am making port refer by-default to the tool hosted port. If we did this I would stop talking about "Diagnostics Server" and instead say "the runtime has a management port" or "the runtime can connect to a tool's management port." We would also get rid of the current DiagnosticMonitorAddress* environment variables. We could support an optional more generalized syntax for DOTNET_MANAGEMENT_PORT allowing lists of ports, specifying if the runtime should connect or listen, and specifying if the runtime should suspend on start:
DOTNET_MANAGEMENT_PORT=\foo\bar\mysocket,connect,suspend;\baz\socket,listen,nosuspend

Reply:
I like this idea.  Essentially all the plumbing for multiple connections has already been done to enable the reverse scenario.  This suggestion would simplify the semantics and make it easier to reason about for customers.

A couple sticky bits come with this though (we discussed some of these offline):
 * This change would allow different connections to request the suspend features.  This would mean that resuming the runtime will become a "stateful" interaction since we would need to know _which_ connection just sent the command and we couldn't resume until all necessary connections have resumed.  This is definitely doable, but we'll need to plumb some state through to account for it.
 * Since this would be using an environment variable with a list, there's the possibility that customers could stomp out some latent configuration.  We would need to add some guidance for tool developers _and_ customers that they should set this variable like `PATH` and only append to it.  This is a lot better than the current situation where _there can only be one_ connection specified and if a customer sets it, they may be stomping a latent configuration.

Overall, I think this is all goodness and shouldn't require _too_ much work (famous last words...).  My one piece of feedback on this is the naming.  Could we use a plural in the env var to hammer in the idea of a list and maybe use `address` instead of `port`?  `address` feels a little less specific, while `port` makes me immediately thing of a TCP port.

Reply:
> Could we use a plural in the env var to hammer in the idea of a list and maybe use `address` instead of `port`? `address` feels a little less specific, while `port` makes me immediately thing of a TCP port.

The word "port" as a noun is used in other areas of computing beside TCP (especially for physical hardware). In general, it just means some endpoint from which you can access (or provide) information when connected.

If you add support, for example, for getting diagnostics information over TCP ports rather than pipes/domain sockets, would you use the same environment variables? Something like "DOTNET_MANAGEMENT_PORT=tcp:7000,listen,suspend"

What about "DOTNET_DIAGNOSTICS_PORTS" and "--diagnostics-ports" or if you do want to avoid "port" then "DOTNET_DIAGNOSTICS_ADDRESSES" and "--diagnostics-addresses"?

I don't think "management" is a good word to use here because you really don't manage anything with the connection (the runtime isn't being managed by this, the application isn't managed by this, etc; you could argue that one manages the diagnostics state of the runtime, but then it would be named something like DOTNET_DIAGNOSTICS_MANAGEMENT_PORT, and then we're getting into way too lengthy naming territory).

Reply:
> Something like "DOTNET_MANAGEMENT_PORT=tcp:7000,listen,suspend"

Yep!

> What about "DOTNET_DIAGNOSTICS_PORTS" and "--diagnostics-ports"

This is fine by me. A couple nit questions to fully scout the naming:
1. Should it be DOTNET_DIAGNOSTIC_PORTS (removed the 'S' in DIAGNOSTICS)? I can't come up with technical reason why, but pronouncing the two words that end in 's' is harder. Is it just me?
2. I could imagine even if the runtime supports multiple ports that tools are under no obligation to do so. Anyone find it confusing if some places used the singular term 'diagnostic port' and others used the plural 'diagnostic ports'? (I don't)
3. Any concern that --diagnostic-port(s) would conflict with other argument names that dotnet-monitor would be likely to use? What are we calling the two dotnet-monitor endpoints where we are serving http traffic?

> if you do want to avoid "port" then "DOTNET_DIAGNOSTICS_ADDRESSES" and "--diagnostics-addresses"?

Personally I prefer port, partly because it makes it a little easier to use "Diagnostic Port" like a proper noun and know what it refers to and partly because it doesn't clash with other service/proxy terminology that is likely to use 'endpoint' or 'address'.

Not sure if the syntax of the list was fully inferable from my examples but this is a little more fleshed out version of that part:
```
Ports := semicolon separated list of 'Port'
Port := address[,optional_tags]
optional_tags := comma separated list of 'tag'. suspend and nosuspend are mutually exclusive, as are listen and connect.
tag := suspend | nosuspend | listen | connect
```
When specified on the runtime env var 'suspend' and 'connect' are the defaults if not specified in the tags list. When specified in the tool argument 'listen' is the default and 'suspend'/'nosuspend' are ignored entirely if the user writes them. Does this syntax make sense? Its easy to change if there are alternate suggestions : )

Reply:
I moved this discussion over to an issue in the runtime repo, so we could track work related to this as we get closer to RC1 of the runtime.

dotnet/runtime#39792